base/inference.jl

```diff
@@ -3405,13 +3408,20 @@ function effect_free(e::ANY, src::CodeInfo, mod::Module, allow_volatile::Bool)
                             # first argument must be immutable to ensure e is affect_free
                             a = ea[2]
                             typ = widenconst(exprtype(a, src, mod))
-                            if !isa(typ, DataType) || typ.mutable || typ.abstract
+                            if !isa(typ, DataType) || typ.mutable || (typ.abstract && !isconstType(typ))
                                 return false
                             end
                         end
                     end
                 end
                 # fall-through
+            elseif is_known_call(e, _apply, src, mod) && length(ea) > 1
+                ft = exprtype(ea[2], src, mod)
+                if isa(ft, Const) && contains_is(_pure_builtins, ft.val)
+                    # fall-through
```

Comment:
It's marked `@pure` though; that should already express that it can be considered effect-free. Couldn't we just ccall `arg_type_tuple` (or write a proper O(n) Julia function) instead of this crazy `O(n^2 * log(n))` implementation though?